### PR TITLE
Refactor READ_QUERY and fix missing arguments

### LINKED ISSUE #37960: Postgres adapter raises error when using a cursor with a read replica
Read replicas are not allowed to create cursors in the Postgres adapter.  This should be allowed as Postgres has no problem doing this with a replicated database.  It looks like the problem is in `active_record/connection_adapters/postgresql/database_statements.rb:70`

```ruby
        READ_QUERY = ActiveRecord::ConnectionAdapters::AbstractAdapter.build_read_query_regexp(
          :begin, :commit, :explain, :select, :set, :show, :release, :savepoint, :rollback, :with
        ) # :nodoc:
        private_constant :READ_QUERY

        def write_query?(sql) # :nodoc:
          !READ_QUERY.match?(sql)
        end
```

The `build_read_query_regexp` should allow the following keywords:
```ruby
:declare,
:fetch,
:move,
:open,
:close
```

Happy to submit a PR if you agree with that approach

### Steps to reproduce
Attempt to declare a cursor when using a read replica (I used `postgresql-cursor`, but you could re-create this with raw SQL as well).  Test case reproducing this below:
```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "activerecord", "6.0.1"
  gem "pg"
  gem "postgresql_cursor"
end

require "active_record"
require "pg"
require "postgresql_cursor"
require "minitest/autorun"
require "logger"



ActiveRecord::Base.logger = Logger.new(STDOUT)

class Post < ActiveRecord::Base
end

class BugTest < Minitest::Test
  def setup
    ActiveRecord::Base.configurations = {
      default_env: {
        primary: {
          adapter: "postgresql",
          database: "ar_test_1"
        },
        replica: {
          adapter: "postgresql",
          database: "ar_test_1",
          replica: true
        }
      }
    }
    ActiveRecord::Base.establish_connection
    ActiveRecord::Base.connects_to(database: { writing: :primary, reading: :replica })
    ActiveRecord::Schema.define do
      create_table :posts, force: true do |t|
      end
    end
  end

  def teardown
    Post.delete_all
  end

  def test_association_stuff
    Post.create!
    ActiveRecord::Base.connected_to(role: :reading) do
      assert_equal([{ "id" => 1 }], Post.all.each_hash.map { |hash| hash })
    end
  end
end
```

### Expected behavior
The reads should execute as expected

### Actual behavior
The following error is raised: `ActiveRecord::ReadOnlyError: Write query attempted while in readonly mode: declare cursor_CURSORNAME no scroll cursor for SELECT...`

### System configuration
**Rails version**:
6.0.1

**Ruby version**:
2.6.5

### PR DESCRIPTION
**Pull request by eileencodes**
* Squashes commits from #37902 so my following changes don't cause a conflict
* Adds a minor refactoring
* Fixes #37960

### CODE REVIEW

**Review by simi — APPROVED** [review]

**Comment by eileencodes** [comment]
I added this in my commit but not here. I didn't add open to the cursor list for postgres because I was getting a syntax error. After researching how it works it seems like DECLARE opens an implicit cursor and I couldn't find any examples of open that actually worked in Rails. If open is needed I'm hoping someone can send a test.

**Comment by eileencodes** [comment]
cc/ @lizdeika @dlangevin

**Comment by simi** [comment]
:thinking: `OPEN` seems to not be actually PostgreSQL SQL command (see list of commands -  https://www.postgresql.org/docs/current/sql-commands.html), but it is part of embedded PL/pgSQL language (https://www.postgresql.org/docs/12/plpgsql-cursors.html#PLPGSQL-CURSOR-OPENING). and Embedded SQL in C (ECPG) (https://www.postgresql.org/docs/12/ecpg-sql-open.html).

TL;DR it seems there's no need to include `OPEN` for PostgreSQL adapter in `READ_QUERY` list.

**Comment by eugeneius** [comment]
Backported to 6-0-stable in https://github.com/rails/rails/commit/99cd347a55fd9bddb458afd003e0f50c5b2ba29a, as this problem just came up again in https://github.com/rails/rails/issues/39932.
